doc/ipmc/IPMC_Data_Plane_HLD.md

```diff
@@ -0,0 +1,472 @@
+
+Listen for NetLink message group type
+
+```text
+RTNLGRP_IPV4_MROUTE
+RTNLGRP_IPV6_MROUTE
+```
+
+Listens for the netlink message type
+
+```text
+RTM_NEWROUTE
+RTM_DELROUTE
+```
+
+The obtained routing information family should be of multicast type
+
+```text
+RTNL_FAMILY_IPMR
+```
+
+#### DB Changes
+
+##### APPL_DB changes
+
+Special note:
+
+1. The multicast route has one more key field than other routes: the source IP address. Therefore, the multicast route cannot share the same TABLE with the original routing entries
+2. When delivering an ipv6 multicast route, assume that both the source and destination IP addresses are ipv6. If the default delimiter of APPL DB ':' is used, the two ipv6 addresses cannot be distinguished by delimiter. Therefore, the delimiter '|' is used here.
+
+MROUTE_TABLE
+
+Producer: mroutesyncd
+
+Consumer: mrouteorch
+
+Description: A table was added to store IP multicast routing data. One entry corresponds to one multicast route.
+
+Schema:
+
+```SQL
+; New table
+; Store IP multicast routing data
+
+key = MROUTE_TABLE:vrf_name|source_ip|dest_ip 
+                                    ; APPL DB usually uses ':' as the separator, but '|' is chosen here.
+                                    ; The reason is that when the ip address is ipv6, using ':' cannot distinguish two ipv6 addresses
+                                    vrf_name    ; The multicast route belongs to the Vrf. If this field is empty, the default global Vrf is used
+                                    source_ip   ; Source IP addresses of multicast routes, for example, IPv4 0.0.0.0(ASM), A.B.C.D(SSM)
+                                    dest_ip     ; Indicates the destination IP address of the multicast route
+
+; field = value
+incoming_group = router_intfs_group  ; IP multicast routes into an interface member group, which can be empty or have multiple members. Each member is separated by commas ','
+outgoing_group = router_intfs_group  ; The IP multicast route outgoing interface member group, which cannot be empty. It can have one or more members. Each member is separated by commas ','
+```
+
+Example:
+
+```bash
+admin@sonic:~$ redis-cli -n 4 hgetall "MROUTE_TABLE|0.0.0.0|225.0.0.1"
+1) "incoming_group"
+2) "Ethernet1"
+3) "outgoing_group"
+4) "Ethernet13,Ethernet15"
+```
+
+```bash
+admin@sonic:~$ redis-cli -n 0 hgetall "MROUTE_TABLE:192.168.1.200|230.0.0.1"
+1) "incoming_group"
+2) "Ethernet49"
+3) "outgoing_group"
+4) "Ethernet39,Ethernet65,Vlan200"
+```
+
+#### SWSS Container
+
+mroutesyncd
+
+- Reads IP multicast routing messages from the kernel and listens for IP multicast routing changes in the kernel
+- Based on netlink messages, the source IP address, destination IP address, inbound interface member, and outbound member of the IP multicast route are resolved and written into the APPL DB
+- During a warm reboot, compare the kernel multicast route and APPL DB data to update the warm reboot data

```

Comment:
Need a mechanism to propagate the IPMC counters in the hardware (SAI_IPMC_ENTRY_ATTR_COUNTER_ID) to the kernel/FRR.

Reply:
In the SAI definition, SAI_IPMC_ENTRY_ATTR_COUNTER_ID means 'packet hits count', which I understand as the statistic for IPMC hit packets.
Why do we need to pass IPMC packet statistics to kernel/FRR? It seems that there is no use case for this at the moment.

Reply:
> Need a mechanism to propagate the IPMC counters in the hardware (SAI_IPMC_ENTRY_ATTR_COUNTER_ID) to the kernel/FRR.

In the SAI definition, SAI_IPMC_ENTRY_ATTR_COUNTER_ID means 'packet hits count', which I understand as the statistic for IPMC hit packets.
Why do we need to pass IPMC packet statistics to kernel/FRR? It seems that there is no use case for this at the moment.